diff --git a/absl/strings/internal/str_format/extension.h b/absl/strings/internal/str_format/extension.h
index 55cbb56..b814d06 100644
--- a/absl/strings/internal/str_format/extension.h
+++ b/absl/strings/internal/str_format/extension.h
@@ -369,9 +369,10 @@ constexpr FormatConversionCharSet FormatConversionCharToConvValue(char conv) {
 }
 
 struct FormatConversionCharSetInternal {
-#define ABSL_INTERNAL_CHAR_SET_CASE(c)         \
-  static constexpr FormatConversionCharSet c = \
-      FormatConversionCharToConvValue(#c[0]);
+#define ABSL_INTERNAL_CHAR_SET_CASE(letter)    \
+  static constexpr FormatConversionCharSet letter = \
+      static_cast<FormatConversionCharSet>(1ULL << \
+          (1 + static_cast<uint8_t>(FormatConversionCharInternal::Enum::letter)));
   ABSL_INTERNAL_CONVERSION_CHARS_EXPAND_(ABSL_INTERNAL_CHAR_SET_CASE, )
 #undef ABSL_INTERNAL_CHAR_SET_CASE
 
