diff --git a/absl/strings/internal/str_format/extension.h b/absl/strings/internal/str_format/extension.h
index 55cbb56..95fca63 100644
--- a/absl/strings/internal/str_format/extension.h
+++ b/absl/strings/internal/str_format/extension.h
@@ -182,7 +182,7 @@ inline std::ostream& operator<<(std::ostream& os, Flags v) {
 struct FormatConversionCharInternal {
   FormatConversionCharInternal() = delete;
 
- private:
+ public:
   // clang-format off
   enum class Enum : uint8_t {
     c, s,                    // text
@@ -369,9 +369,10 @@ constexpr FormatConversionCharSet FormatConversionCharToConvValue(char conv) {
 }
 
 struct FormatConversionCharSetInternal {
-#define ABSL_INTERNAL_CHAR_SET_CASE(c)         \
-  static constexpr FormatConversionCharSet c = \
-      FormatConversionCharToConvValue(#c[0]);
+#define ABSL_INTERNAL_CHAR_SET_CASE(letter)    \
+  static constexpr FormatConversionCharSet letter = \
+      static_cast<FormatConversionCharSet>(1ULL << \
+          (1 + static_cast<uint8_t>(FormatConversionCharInternal::Enum::letter)));
   ABSL_INTERNAL_CONVERSION_CHARS_EXPAND_(ABSL_INTERNAL_CHAR_SET_CASE, )
 #undef ABSL_INTERNAL_CHAR_SET_CASE
 
